app/views/sign_up/recovery_codes/_recovery_page_controller.js.erb

```diff
@@ -0,0 +1,96 @@
+var modalSelector = '#personal-key-confirm';
+var modal = new LoginGov.Modal({el: modalSelector });
+
+var recoveryWords = [].slice.call(document.querySelectorAll('[data-recovery]'));
+var formEl = document.getElementById('confirm-key');
+var inputs = [].slice.call(formEl.elements).filter(function(el) { return el.type === 'text'});
+var modalTrigger = document.querySelector('[data-toggle="modal"]');
+var modalDismiss = document.querySelector('[data-dismiss="personal-key-confirm"]')
+
+var reminderEl = document.getElementById('recovery-code-reminder');
+
+var isInvalidForm = false;
+var alertFallback = '<div class="alert alert-alert" role="alert" id="recovery-code-alert">' +
+                    "<%= t('users.recovery_code.confirmation_error') %>" +
```

Comment:
To me, it's not a Rails app thing, it's an XSS thing. Building HTML in a string in JS like that is very easy to break out of.

Personally I like looking at the HTML like that since it's easier to see what the structure is than from a `document.createElement`, so I would:


1. Add the HTML outline to the page in a normal .html.erb (or .slim) template, but give it a `hidden` class. If needed, set the error content later, but it looks like a static error message so I think we'll be fine

2. Show the HTML as needed in the JS